receiver/nvmlreceiver/client.go

```diff
@@ -0,0 +1,276 @@
+	return
+}
+
+func discoverDevices(logger *zap.Logger) ([]nvml.Device, []string, error) {
+	count, ret := nvml.DeviceGetCount()
+	if ret != nvml.SUCCESS {
+		return nil, nil, fmt.Errorf("Unable to get Nvidia device count on '%v'", nvml.ErrorString(ret))
+	}
+
+	devices := make([]nvml.Device, count)
+	names := make([]string, count)
+	for i := 0; i < count; i++ {
+		devices[i], ret = nvml.DeviceGetHandleByIndex(i)
+		if ret != nvml.SUCCESS {
+			return nil, nil, fmt.Errorf("Unable to get Nvidia device at index %d on '%v'", i, nvml.ErrorString(ret))
+		}
+
+		uuid, ret := devices[i].GetUUID()
+		if ret != nvml.SUCCESS {
+			return nil, nil, fmt.Errorf("Unable to get UUID of Nvidia device %d on '%v'", i, nvml.ErrorString(ret))
+		}
+
+		names[i], ret = devices[i].GetName()
+		if ret != nvml.SUCCESS {
+			return nil, nil, fmt.Errorf("Unable to get name of Nvidia device %d on '%v'", i, nvml.ErrorString(ret))
+		}
+
+		logger.Sugar().Infof("Discovered Nvidia device %d of model %s with UUID %s", i, names[i], uuid)
+
+		currMode, _, ret := devices[i].GetMigMode()
+		if ret != nvml.SUCCESS {
+			logger.Sugar().Warnf("Unable to query MIG mode for Nvidia device %d", i)
+			continue
+		}
+		if currMode == nvml.DEVICE_MIG_ENABLE {
+			logger.Sugar().Warnf("Nvidia device %d has MIG enabled. GPU utilization queries may not be supported", i)
+		}
+	}
+
+	return devices, names, nil
+}
+
+func (client *nvmlClient) cleanup() {
+	if client.handleCleanup != nil {
+		client.handleCleanup()
+	}
+	if !client.disable {
+		client.logger.Info("Shutdown Nvidia Management Library client")
+	}
+}
+
+func (client *nvmlClient) getDeviceModelName(gpuID uint) string {
+	return client.devicesModelName[gpuID]
+}
+
+func (client *nvmlClient) collectDeviceMetrics() ([]nvmlMetric, error) {
+	// not strictly needed since len(client.devices) = 0; but, safer
+	if client.disable {
+		return nil, nil
+	}
+
+	deviceMetrics := client.collectDeviceUtilization()
+	deviceMetrics = append(deviceMetrics, client.collectDeviceMemoryInfo()...)
+	return deviceMetrics, nil
+}
+
+func (client *nvmlClient) collectDeviceUtilization() []nvmlMetric {
+	deviceMetrics := make([]nvmlMetric, 0, len(client.devices))
+
+	gpuUtil := nvmlMetric{name: "nvml.gpu.utilization"}
+
+	for idx, device := range client.devices {
+		mean, err := client.getAverageGpuUtilizationSinceLastQuery(device)
+		if err != nil {
+			client.issueWarningForFailedQueryUptoThreshold(idx, gpuUtil.name, err.Error())
+			continue
+		}
+
+		gpuUtil.gpuID = uint(idx)
+		gpuUtil.time = time.Now()
+		gpuUtil.setFloat64(mean)
+		deviceMetrics = append(deviceMetrics, gpuUtil)
+		client.logger.Debugf("Nvidia device %d has GPU utilization of %.1f%%", idx, 100.0*gpuUtil.asFloat64())
+	}
+
+	return deviceMetrics
+}
+
+func (client *nvmlClient) getAverageGpuUtilizationSinceLastQuery(device nvml.Device) (float64, error) {
+	nvmlType, samples, ret := nvmlDeviceGetSamples(device, nvml.GPU_UTILIZATION_SAMPLES, client.deviceToLastSeenTimestamp[device])
```

Comment:
Good question. So in the [documentation](https://docs.nvidia.com/deploy/nvml-api/group__nvmlDeviceQueries.html#group__nvmlDeviceQueries_1g540824faa6cef45500e0d1dc2f50b321) you'll see nvmlDeviceGetUtilizationRates returns a single [nvmlUtilization_t](https://docs.nvidia.com/deploy/nvml-api/structnvmlUtilization__t.html#structnvmlUtilization__t) struct. And it states "Each sample period may be between 1 second and 1/6 second, depending on the product being queried." We are only scraping every 60 seconds, so one sample representing the last 1/6 to 1 second won't be a sufficient average for us. On the other hand  nvmlDeviceGetSamples returns all the samples still in cache (about 100).

Reply:
Also discussed this offline. Thanks for explaining this to me. 
Another note from the discussion is that memory utilization normally changes less within a sampling period, so it would be sufficient to use `DeviceGetMemoryInfo` for memory utilization.